controllers/storagecluster/consumer.go

```diff
@@ -0,0 +1,146 @@
+package storagecluster
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	providerClient "github.com/red-hat-storage/ocs-operator/services/provider/client"
+)
+
+var (
+	externalOCSResources map[types.UID][]ExternalResource
+)
+
+type ocsConsumer struct{}
+
+func (o *ocsConsumer) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+	if !(instance.Spec.ExternalStorage.Enable && instance.Spec.ExternalStorage.StorageProviderKind == ocsv1.KindOCS) {
+		return nil
+	}
+
+	if instance.Status.ExternalStorage.ConsumerID == "" {
+		return onBoardConsumer(r, instance)
+	}
+
+	if !instance.Status.ExternalStorage.GrantedCapacity.Equal(*instance.Spec.ExternalStorage.RequestedCapacity) {
+		return upscaleOrDownscaleConsumer(r, instance)
+	}
+
+	return nil
+}
+
+func (o *ocsConsumer) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+	if !(instance.Spec.ExternalStorage.Enable && instance.Spec.ExternalStorage.StorageProviderKind == ocsv1.KindOCS) {
+		return nil
+	}
+
+	return offBoardConsumer(r, instance)
+}
+
+func onBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	consumerClient, err := getConsumerClient(instance)
+	if err != nil {
+		return err
+	}
+
+	capacity := func(instance *ocsv1.StorageCluster) string {
+		if instance.Spec.ExternalStorage.RequestedCapacity == nil {
+			q := resource.MustParse("1Ti")
+			instance.Spec.ExternalStorage.RequestedCapacity = &q
+		}
+		return instance.Spec.ExternalStorage.RequestedCapacity.String()
+	}(instance)
+
+	// TODO: think about naming convention
+	response, err := consumerClient.OnboardConsumer("my-ocs", capacity)
+	if err != nil {
+		requeue := r.handleOnBoardConsumerErrors(instance, err)
+		if !requeue {
+			return nil
+		}
+		return err
+	}
+
+	if response.StorageConsumerUUID == "" || response.GrantedCapacity == "" {
+		return fmt.Errorf("response is empty")
+	}
+
+	instance.Status.ExternalStorage.ConsumerID = response.StorageConsumerUUID
+	instance.Status.ExternalStorage.GrantedCapacity = resource.MustParse(response.GrantedCapacity)
+
+	return nil
+}
+
+func offBoardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	return nil
+}
+
+func upscaleOrDownscaleConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	return nil
+}
+
+func getExternalConfigFromProvider(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) ([]ExternalResource, error) {
+
+	if externalOCSResources[instance.UID] != nil {
+		return externalOCSResources[instance.UID], nil
+	}
+
+	consumerClient, err := getConsumerClient(instance)
+	if err != nil {
+		return nil, err
+	}
+
+	consumerID := instance.Status.ExternalStorage.ConsumerID
```

Comment:
Please use instance.Status.ExternalStorage.ConsumerID and remove the consumer ID variable that is not used

Reply:
done